addOns/paramminer/src/main/java/org/zaproxy/addon/paramminer/ParamGuesserScanController.java

```diff
@@ -0,0 +1,254 @@
+    }
+
+    @Override
+    public void stopScan(int id) {
+        paramGuesserScanLock.lock();
+        try {
+            if (this.paramGuesserScanMap.containsKey(id)) {
+                this.paramGuesserScanMap.get(id).stopScan();
+                updateScansPanel();
+            }
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void stopAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            paramGuesserScanMap.values().forEach(GuesserScan::resumeScan);
+            updateScansPanel();
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void pauseScan(int id) {
+        paramGuesserScanLock.lock();
+        try {
+            if (this.paramGuesserScanMap.containsKey(id)) {
+                this.paramGuesserScanMap.get(id).pauseScan();
+                updateScansPanel();
+            }
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void pauseAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            paramGuesserScanMap.values().forEach(GuesserScan::pauseScan);
+            updateScansPanel();
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void resumeScan(int id) {
+        paramGuesserScanLock.lock();
+        try {
+            if (this.paramGuesserScanMap.containsKey(id)) {
+                this.paramGuesserScanMap.get(id).resumeScan();
+                updateScansPanel();
+            }
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void resumeAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            paramGuesserScanMap.values().forEach(GuesserScan::resumeScan);
+            updateScansPanel();
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public GuesserScan removeScan(int id) {
+        paramGuesserScanLock.lock();
+        try {
+            if (!this.paramGuesserScanMap.containsKey(id)) {
+                return null;
+            }
+            GuesserScan currentScan = this.paramGuesserScanMap.get(id);
+            currentScan.stopScan();
+            currentScan.clear();
+            this.paramGuesserScanMap.remove(id);
+            return currentScan;
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public int removeAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            int count = 0;
+            for (Iterator<GuesserScan> it = paramGuesserScanMap.values().iterator();
+                    it.hasNext(); ) {
+                GuesserScan ascan = it.next();
+                ascan.stopScan();
+                ascan.clear();
+                it.remove();
+                count++;
+            }
```

Comment:
Couldn't the map just be streamed or couldn't this use forEach/advanced loop instead of creating an iterator?

Reply:
It could.

Reply:
ascan to scan done :)